### PR TITLE
EasyPry airlocks for arrivals. Now also prying refactor I guess

### PR DESCRIPTION
**Pull request by nikthechampiongr**
## About the PR

This PR adds airlock prototypes that can be opened without any tools while they are unpowered. This does not change any existing doors, they would have to be added to maps by mappers. I am not quite sure the prototypes I made are enough, or too many as I am not quite sure what kind of airlocks will be used for this.

In the process of fulfilling maintainer requests for this, it turned into a refactor of the way doors are pried. Now it's dependent on its own component and system.
## Why / Balance

This is mostly to resolve the current issues with the arrivals shuttle where a power outage, and missing crowbars from the shuttle leads to people being stranded in space once it leaves. This will allow players to not be stuck outside the station before they can even begin to start playing.

## Technical details

A door can have a pryunpowered component. The door system will make a call to the prying system if it fails to open a door. If the door has the pryunpowered component, then the system will try to start prying the door going through normal checks, effectively acting like a tool with a doorprying component was used.

The prying component marks an entity as able to pry an entity. If the datafield PryPowered is set to true, then it will be able to pry a door even if it is still powered, all other checks still apply. The prying component can be disabled to stop a tool from prying

When a prying attempt is made, the prying system checks that the entity has the prying component (or if it's an easypry attempt see above), and sends out an event which other systems can use to cancel the prying attempt. If it doesn't get cancelled then an event is sent out for systems to add multipliers to the prying time or change the base time(for example if the door is powered): After that the doafter starts with any modifiers applied to it by the previously mentioned event, and modifiers the entity with the prying component may have. If it finishes successfully then the prying sound is played(set by the UseSound datafield, defaults to the crowbar sound) and the entity is pried.
## Media

https://github.com/space-wizards/space-station-14/assets/32041239/4ef8ff35-656d-488d-b900-6e4c4e68c266

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes

A tool being able to pry doors no longer solely relies on the prying tool quality. In order for it to be able to pry doors it will need the DoorPrying component. Entities like xenos or zombies can also have the doorprying component with the requiresComponent datafield being set to false which allows them to pry open doors without tools.

**Changelog**

:cl: 
- add: Mappers can now place airlocks that can be pried open while unpowered without tools.

### CODE REVIEW

**Comment by EmoGarbage404** [comment]
from a UX perspective, it's really strange that these random doors on the shuttle are the only ones that can be pried with bare hands. I feel like we should just be making the doors forced open or something like that. A lot of players could miss the easy prying functionality just because it's fairly nonstandard behavior that isn't indicated well.

**Comment by nikthechampiongr** [comment]
> from a UX perspective, it's really strange that these random doors on the shuttle are the only ones that can be pried with bare hands.
Fair enough, this is something I lifted from ss13(Don't quote me on that I Just remember that arrivals airlocks could just do this). I could add something in the description of the airlocks that could be seen when you examine it.

> I feel like we should just be making the doors forced open or something like that.
Wouldn't this be even more strange? Just airlocks randomly being opened once power goes out, not to mention that these doors would be open to space.

**Comment by deltanedas** [comment]
theyre fanned so its fine
also fnaf at freddy does it so

**Comment by nikthechampiongr** [comment]
> theyre fanned so its fine 

Still looks very wrong in my opinion for doors to space to open up once power goes out, and it's even weirder when the atmosphere just stays in even tho that fan has also technically lost power. 

> also fnaf at freddy does it so

Not sure how to respond to that.

**Comment by Jezithyr** [comment]
Maybe update the description to say it has an emergency release? Similar kinda like the doors on subway cars, where if the power goes out you can pump a level to open/close the doors. I do agree with @EmoGarbage404 that the doors need to be obviously different from the regular doors on the station, which might require a sprite update.

**Comment by nikthechampiongr** [comment]
> Maybe update the description to say it has an emergency release? Similar kinda like the doors on subway cars, where if the power goes out you can pump a lever to open/close the doors. I do agree with @EmoGarbage404 that the doors need to be obviously different from the regular doors on the station, which might require a sprite update.

I updated their descriptions to say that they say they got a handle labelled "PULL TO OPEN". Not sure if that's the best way to describe it tho. On sprites imma be honest, I'm creatively bankrupt so I'm not sure if I could pull off good looking alternative sprites.

**Comment by nikthechampiongr** [comment]
Thinking about it more, it should probably say lever instead of handle, I'll update it tomorrow

**Comment by hord-brayden** [comment]
Unsure if it's been stated, but there's an IRL precedence for this. Submarine doors as well have this feature for brown outs/powerless scenarios. It's usually a wheel that must be spun vigorously to slowly open the bolts.

The arrivals shuttle in general is so scuffed that it just plops you there. I think this is a great fix.

**Comment by Partmedia** [comment]
Thank you for your contribution. I'm conceptually in favor of having this, particularly given the real life precedent. The problem of the arrivals shuttle was resolved by merging #17189 but there may still be other uses for this door.

Unfortunately I haven't been able to take a look at the code yet.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]
Just make it a door and remove bumpopen and it does the exact same thing.

**Comment by nikthechampiongr** [comment]
> Just make it a door and remove bumpopen and it does the exact same thing.

Hey are you sure I should just do it like this? At least to my understanding this would just make the door immediately open regardless of if  it is powered or not since it will probably be used on airlocks. Currently it has a doafter for prying it open while it is unpowered, do you want me to scrap that?

**Comment by nikthechampiongr** [comment]
> Hey are you sure I should just do it like this? At least to my understanding this would just make the door immediately open regardless of if it is powered or not since it will probably be used on airlocks. Currently it has a doafter for prying it open while it is unpowered, do you want me to scrap that?

To be clearer on what I mean by this, currently when trying to open a door it fires off a BeforeDoorOpenedEvent which is used by the airlocksystem to stop opening unpowered airlocks. I could make it so that if the airlock is marked as easypry it does not cancel the event but it would have the effect I mentioned above. I could add back the doafter by making the doorsystem check this as well but this would be mixing logic between systems which I am not sure would be good.

**Comment by metalgearsloth** [comment]
> > Hey are you sure I should just do it like this? At least to my understanding this would just make the door immediately open regardless of if it is powered or not since it will probably be used on airlocks. Currently it has a doafter for prying it open while it is unpowered, do you want me to scrap that?
> 
> To be clearer on what I mean by this, currently when trying to open a door it fires off a BeforeDoorOpenedEvent which is used by the airlocksystem to stop opening unpowered airlocks. I could make it so that if the airlock is marked as easypry it does not cancel the event but it would have the effect I mentioned above. I could add back the doafter by making the doorsystem check this as well but this would be mixing logic between systems which I am not sure would be good.

Prying shouldn't be getting handled by both tools and a dedicated system, it should just be 1 dedicated system. Alternatively moving all prying to 1 component is probably best.

**Comment by hord-brayden** [comment]
> > > Hey are you sure I should just do it like this? At least to my understanding this would just make the door immediately open regardless of if it is powered or not since it will probably be used on airlocks. Currently it has a doafter for prying it open while it is unpowered, do you want me to scrap that?
> > 
> > 
> > To be clearer on what I mean by this, currently when trying to open a door it fires off a BeforeDoorOpenedEvent which is used by the airlocksystem to stop opening unpowered airlocks. I could make it so that if the airlock is marked as easypry it does not cancel the event but it would have the effect I mentioned above. I could add back the doafter by making the doorsystem check this as well but this would be mixing logic between systems which I am not sure would be good.
> 
> Prying shouldn't be getting handled by both tools and a dedicated system, it should just be 1 dedicated system. Alternatively moving all prying to 1 component is probably best.

Just spitballing here:

Unsure if zombie prying is housed within its own system, but assuming it is: if we were to nix that, and within prying make some doors pryable without tools by all races, zombies can extend that, everything should be neatly packaged - or no?

**Comment by nikthechampiongr** [comment]
> Unsure if zombie prying is housed within its own system, but assuming it is: if we were to nix that, and within prying make some doors pryable without tools by all races, zombies can extend that, everything should be neatly packaged - or no?
I'll look if I can do that, I am not really familiar with how zombies do that but yea it may be a good fit for the prying component/system

**Comment by nikthechampiongr** [comment]
> I'll look if I can do that, I am not really familiar with how zombies do that but yea it may be a good fit for the prying component/system

I figured it out. It's using the GetVerbsEvent in the door system and I think it applies for all mobs able to pry stuff like xenos. If the mob can pry stuff then it also makes it so most checks are skipped like whether the door is powered or bolted. Oddly it still checks if the door is welded due to the way it's coded. I'll just make it so that event just calls into the prying system. This will make it so it actually does checks except for whether the door is powered for those cases due to the way I've coded it so far tho. This does also change zombie behavior which is kinda out of scope of this pr, but considering I've been asked to create an entire new system and component for prying I think I'll need to redo the entire pr name and description anyways.

**Comment by nikthechampiongr** [comment]
Well that's unfortunate, anyways this is what I have in having prying be its own component and system. It does have a glitch at the moment with the door opening sound being triggered twice so even if it didn't have merge conflicts it shouldn't be merged at the moment

File: Content.Server/Doors/Systems/DoorSystem.cs [hunk 1]
```diff
@@ -49,7 +53,9 @@ protected override void OnActivate(EntityUid uid, DoorComponent door, ActivateIn
         if (args.Handled || !door.ClickOpen)
             return;
 
-        TryToggleDoor(uid, door, args.User);
+        if (!TryToggleDoor(uid, door, args.User))
+            _pryingSystem.TryPry(uid, args.User, door, out _);
```
[the remark this answers is not in the document]
Comment (nikthechampiongr) [reply Content.Server/Doors/Systems/DoorSystem.cs:57]: If I do that then doors that can be pried without tools will give you a "The powered motors block your efforts" message.

**Comment by nikthechampiongr** [comment]
The issue with the sound playing twice currently remains. For some reason calling opendoor or closedoor from the shared door system calls the server doorsystem at the same time and I am not sure how to fix it.

**Comment by nikthechampiongr** [comment]
Ok I've discovered this doesn't fully address tools like the jaws of life due to it having multiple state. This may take a bit.

**Comment by nikthechampiongr** [comment]
Yea at this point I am not sure extracting prying out of the tool system can work well. It's extremely coupled.

**Comment by nikthechampiongr** [comment]
Anyways the solution I came up with for the problem with the jaws of life is to just fire an event that enables or disables the component when a switch happens. I really don't like it but it's the  only way to make it so that prying doors doesn't depend on the tool component as I was asked.

**Comment by nikthechampiongr** [comment]
Nevermind I found a better solution in the tilecomponent, will just use that

**Comment by nikthechampiongr** [comment]
Alright then, the only remaining issue is the sound of the door opening and closing plays twice because for a reason I cannot comprehend, a call to the shared door system will also call the same thing in the server system at the exact same time.

**Comment by nikthechampiongr** [comment]
I removed the playsound call on the shared door system. It appears that everything works just fine tho. I tested out opening and closing doors, and using the door remotes. Sound seemed to play just fine.

**Comment by nikthechampiongr** [comment]
Upon further testing, the door closing sound broke. However now the same sound issue with the closing sound being triggered twice is back, now only when prying a door closed. I have absolutely no idea why, as this time there isn't even the same method on the server door component. I am at an absolute loss and have no idea how to fix it.

**Comment by nikthechampiongr** [comment]
what the fuck

**Comment by nikthechampiongr** [comment]
Well I guess it kinda works again, only the tools used for prying now make no sound again for some reason... I am too tired right now to deal with my code randomly breaking due to a seemingly unrelated change in the rest of the codebase, or to deal with a random modification to the toolbox module that somehow madeit into a commit and also does not show up

**Comment by nikthechampiongr** [comment]
ok now apparently it fails the same way on my machine like it did on the tests. Why?

File: Content.Shared/Doors/Systems/PryingSystem.cs [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+using Robust.Shared.Serialization;
+using Content.Shared.Administration.Logs;
+using Content.Shared.Database;
+using Content.Shared.Doors.Components;
+using Content.Shared.Doors.Systems;
+using Content.Shared.Tools.Components;
+
+namespace Content.Shared.Doors.Prying.Systems;
+public class DoorPryingSystem : EntitySystem
+{
```
Comment (metalgearsloth) [inline Content.Shared/Doors/Systems/PryingSystem.cs:12]: Please make this sealed as per the warning and add a space with the namespace.

This should also be called PryingSystem not DoorPryingSystem as it shouldn't be hardcoded to doors.

File: Content.Shared/Doors/Systems/PryingSystem.cs [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    ///     Pry open a door.
+    /// </summary>
+    public bool TryPry(EntityUid target, EntityUid user, DoorComponent door, out DoAfterId? id, EntityUid tool)
+    {
+        id = null;
+
+        DoorPryingComponent? comp = null;
+        if (!Resolve(tool, ref comp))
+            return false;
+
```
Comment (metalgearsloth) [inline Content.Shared/Doors/Systems/PryingSystem.cs:35]: Prying shouldn't be tied to doors here, it should just be generic.

File: Content.Shared/Doors/Systems/PryingSystem.cs [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+        if (comp.NeedsComponent)
+        {
+            ToolComponent? tComp = null;
+            if (!Resolve(tool, ref tComp) || !tComp.Qualities.Contains("Prying"))
+                return false;
+        }
```
Comment (metalgearsloth) [inline Content.Shared/Doors/Systems/PryingSystem.cs:41]: Just make the tool quality the pryingcomponent.

File: Content.Shared/Doors/Systems/PryingSystem.cs [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        SubscribeLocalEvent<DoorComponent, DoorPryDoAfterEvent>(OnDoAfter);
+    }
+
+    /// <summary>
+    ///     Pry open a door.
+    /// </summary>
```
Comment (metalgearsloth) [inline Content.Shared/Doors/Systems/PryingSystem.cs:27]:
```suggestion
    /// <summary>
    ///     Attempt to pry open a door.
    /// </summary>
```

File: Content.Shared/Doors/Systems/PryingSystem.cs [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+        RaiseLocalEvent(target, canev, false);
+
+        if (canev.Cancelled)
+            return false;
+        return true;
+    }
+
+    void StartPry(EntityUid target, EntityUid user, DoorComponent door, EntityUid? tool, float toolModifier, out DoAfterId? id)
+    {
+        var modEv = new DoorGetPryTimeModifierEvent(user);
```
Comment (metalgearsloth) [inline Content.Shared/Doors/Systems/PryingSystem.cs:82]: Private methods should be marked as such.

File: Content.Shared/Doors/Systems/PryingSystem.cs [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+    /// </summary>
+    public bool TryPry(EntityUid target, EntityUid user, DoorComponent door, out DoAfterId? id)
+    {
+        id = null;
+
+        if (!TryComp<PryUnpoweredComponent>(target, out _))
+            return false;
+
+        if (!CanPry(target, user, false))
+            return true;
```
Comment (metalgearsloth) [inline Content.Shared/Doors/Systems/PryingSystem.cs:63]: PryUnpowered should be in CanPry and not separate.

File: Content.Shared/Doors/Systems/PryingSystem.cs [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+        }
+
+        StartPry(target, user, door, tool, comp.SpeedModifier, out id);
+        return true;
+    }
+
+    /// <summary>
+    ///     Pry open a door.
+    /// </summary>
+    public bool TryPry(EntityUid target, EntityUid user, DoorComponent door, out DoAfterId? id)
```
Comment (metalgearsloth) [inline Content.Shared/Doors/Systems/PryingSystem.cs:55]: As above.

Comment (metalgearsloth) [inline Content.Shared/Doors/Systems/PryingSystem.cs:55]: Is this id supposed to be marked as notnullwhen(true)

File: Content.Shared/Doors/Systems/PryingSystem.cs [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+    {
+        if (args.Cancelled)
+            return;
+        if (args.Target is null)
+            return;
+
+        DoorPryingComponent? comp = null;
+
+        if (args.Used != null && Resolve(args.Used.Value, ref comp))
+            _audioSystem.PlayPredicted(comp.UseSound, args.Used.Value, args.User, comp.UseSound.Params.WithVariation(0.175f).AddVolume(-5f));
```
Comment (metalgearsloth) [inline Content.Shared/Doors/Systems/PryingSystem.cs:113]: Passing in manual audio params is deprecated and the sound should be specifying it.

File: Content.Shared/Doors/Components/PryingComponent.cs [hunk 10]
```diff
@@ -0,0 +1,66 @@
+using Robust.Shared.Audio;
+using Robust.Shared.GameStates;
+
+namespace Content.Shared.Doors.Prying.Components;
+
+[RegisterComponent, NetworkedComponent]
+public sealed partial class DoorPryingComponent : Component
```
Comment (metalgearsloth) [inline Content.Shared/Doors/Components/PryingComponent.cs:7]: This should be renamed to PryingComponent.

File: Content.Shared/Doors/Components/PryUnpoweredComponent.cs [hunk 11]
```diff
@@ -0,0 +1,8 @@
+namespace Content.Shared.Doors.Prying.Components;
+
+///<summary>
+/// Applied to doors that can be pried open without tools while unpowered
+/// </summary>
+[RegisterComponent]
+public sealed partial class PryUnpoweredComponent : Component
```
Comment (metalgearsloth) [inline Content.Shared/Doors/Components/PryUnpoweredComponent.cs:7]: This should be networked.

File: Content.Shared/Doors/Components/PryUnpoweredComponent.cs [hunk 12]
```diff
@@ -0,0 +1,8 @@
+namespace Content.Shared.Doors.Prying.Components;
+
+///<summary>
+/// Applied to doors that can be pried open without tools while unpowered
```
Comment (metalgearsloth) [inline Content.Shared/Doors/Components/PryUnpoweredComponent.cs:4]:
```suggestion
/// Applied to entities that can be pried open without pryingcomponent while unpowered.
```

Except use the <see cref> for pryingcomponent.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

File: Content.Shared/Doors/Systems/PryingSystem.cs (hunk 8 above)
Comment (nikthechampiongr) [reply Content.Shared/Doors/Systems/PryingSystem.cs:55]: No, I was just missing some documentation that stated that it returned true to mark an interactusingevent as handled even if a doafter had not occurred in some cases.

**Review by metalgearsloth — APPROVED** [review]

**Comment by Partmedia** [comment]
I think there may be a defect in easy pry. For a door that is currently powered, but has PryUnpoweredComponent, prying appears to bypass PryPowered check.

**Comment by nikthechampiongr** [comment]
> I think there may be a defect in easy pry. For a door that is currently powered, but has PryUnpoweredComponent, prying appears to bypass PryPowered check.

Please make an issue and I'll fix it. It should just be a 1 line fix.

**Pull request merged**